### PR TITLE
Migrate Session class from Java to Kotlin (part 4)

### LINKED ISSUE #631: Migrate Session class from Java to Kotlin
# Current state
- The [`Session` class](https://github.com/EventFahrplan/EventFahrplan/blob/7854ff7f1555857c86e1ae4f5c7ab9098055ea30/app/src/main/java/nerd/tuxmobil/fahrplan/congress/models/Session.java) in the `app` module is written in Java.
- As the central data holder in the UI layer its usage is wide-spread and critical to all app features.
- Fields of the class are public, hence can be modified from external.
- The class contains static functions which are contextual related to the `Session` class.
- Not all field are part of the `equals` and `hashCode` methods.

# Target state
- The `Session` class in the `app` module is written in Kotlin as a `data class`.
- Properties of the class are initialized at construction time and cannot be modified. This makes `Session` instances immutable.
- Static functions are moved into separate well named files/classes.
- Not all field are part of the `equals` and `hashCode` functions implicit `data class`. This might be addressed in subsequent issues / merge requests.

# Motivation
- Mutable `Session` instances which miss the benefits of a Kotlin `data class` make development uncomfortable.
- Writing tests with `Session` Java instances is cumbersome and painful. Writing test should be fun!

# Approach
- The migration happens step-wise to facilitate small, easy to comprehend commits.
- Clean-up happens before larger, and functional changes.
- Tests are added before a change to prevent introducing bugs.

# Related
- https://github.com/EventFahrplan/EventFahrplan/issues/448

### PR DESCRIPTION
**Pull request by johnjohndoe**
# Description
- Ease maintainence of `SessionTest` & `ScheduleChangesTest` by using parameterized tests.
- Migrate `:app:Session` class from Java to Kotlin data class. :rocket: 

# Successfully tested on
with `ccc37c3` flavor, `debug` build
- :heavy_check_mark: Samsung Galaxy Tab S7 FE, SM-T733 (tablet), Android 14 (API 34)

# Related
- https://github.com/EventFahrplan/EventFahrplan/issues/631
- https://github.com/EventFahrplan/EventFahrplan/pull/632
- https://github.com/EventFahrplan/EventFahrplan/pull/633
- https://github.com/EventFahrplan/EventFahrplan/pull/634

---

Resolves #631

### CODE REVIEW

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/models/Session.kt [hunk 1]
```diff
@@ -1,287 +1,162 @@
...
+     *
+     * Don't use in [Conference.ofSessions] as long as [relStartTime] is supported.
+     * See: [5a402](https://github.com/EventFahrplan/EventFahrplan/commit/5a4022b00434700274a824cc63f6d54a18b06fac)
      */
-    public Moment getStartsAt() {
-        if (dateUTC <= 0) {
-            throw new IllegalArgumentException("Field 'dateUTC' must be more than 0.");
+    val startsAt: Moment
+        get() {
+            require(dateUTC > 0) { "Field 'dateUTC' must be more than 0." }
```
Comment (cketti) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/models/Session.kt:79]: `require()` throws an `IllegalArgumentException`, but there is no argument. Use `check()` which throws `IllegalStateException` instead.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/models/SessionTest.kt [hunk 2]
```diff
@@ -120,7 +121,7 @@ class SessionTest {
     @Test
     fun `copy creates a new instance`() {
```
Comment (cketti) [inline app/src/test/java/nerd/tuxmobil/fahrplan/congress/models/SessionTest.kt:122]: Now that `Session` is a data class, I'd remove this test. If you keep it, you should change to `assertThat(session2)…`

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/repositories/AppRepositorySessionsTest.kt [hunk 3]
```diff
@@ -47,67 +47,89 @@ class AppRepositorySessionsTest {
...
+            sessionId = "3001",
+            changedIsCanceled = false,
+        )
+
+        private val SESSION_3002 = createSession(
+            sessionId = "3002",
+            changedIsCanceled = true,
+        )
+
+        private fun createSession(
```
Comment (cketti) [inline app/src/test/java/nerd/tuxmobil/fahrplan/congress/repositories/AppRepositorySessionsTest.kt:119]: This feels like a needless indirection now. The `Session` constructor could be used directly.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/sharing/SimpleSessionFormatTest.kt [hunk 4]
```diff
@@ -26,42 +26,44 @@ class SimpleSessionFormatTest {
...
-        slug = "U9SD23"
-    }
+    private val session1 = Session(sessionId = "S1",
+        title = "A talk which changes your life",
+        roomName = "Yellow pavilion",
+        dateText = "2019-12-27T11:00:00+01:00",
+        dateUTC = DateParser.parseDateTime("2019-12-27T11:00:00+01:00"),
+        url = "https://example.com/2019/LD3FX9.html",
+        slug = "LD3FX9",
+        )
```
Comment (cketti) [inline app/src/test/java/nerd/tuxmobil/fahrplan/congress/sharing/SimpleSessionFormatTest.kt:36]: Inconsistent indentation.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/sharing/SimpleSessionFormatTest.kt [hunk 5]
```diff
@@ -26,42 +26,44 @@ class SimpleSessionFormatTest {
...
+
+    private val session2 = Session(sessionId = "S2",
+        title = "The most boring workshop ever",
+        roomName = "Dark cellar",
+        dateText = "2019-12-28T17:00:00+01:00",
+        dateUTC = DateParser.parseDateTime("2019-12-28T17:00:00+01:00"),
+        url = "https://example.com/2019/U28VSA.html",
+        slug = "U28VSA",
+            )
```
Comment (cketti) [inline app/src/test/java/nerd/tuxmobil/fahrplan/congress/sharing/SimpleSessionFormatTest.kt:45]: Inconsistent indentation.

Have you considered automatically enforcing a code style? We're using [spotless](https://github.com/diffplug/spotless) (which uses ktlint for Kotlin files).

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/serialization/ScheduleChangesTest.kt
Comment (cketti) [inline app/src/test/java/nerd/tuxmobil/fahrplan/congress/serialization/ScheduleChangesTest.kt:1]: I don't think this is an improvement. I'd keep the previous version.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/models/Session.kt [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        changedDuration = false,
+        changedSpeakers = false,
+        changedLanguage = false,
+        changedRecordingOptOut = false,
+        changedTrack = false,
+    )
+
+}
+
+private data class EssentialSession(
```
Comment (cketti) [inline app/src/main/java/nerd/tuxmobil/fahrplan/congress/models/Session.kt:122]: For future consideration: This is a good indicator that `Session` is doing too much and should be split apart.

The `changed*` properties are rarely necessary and could live in a separate class

```kotlin
data class ChangedSession(
    val session: Session,

    val changedTitle: Boolean = false,
    val changedSubtitle: Boolean = false,
    val changedRoomName: Boolean = false,
    val changedDayIndex: Boolean = false,
    val changedStartTime: Boolean = false,
    val changedDuration: Boolean = false,
    val changedSpeakers: Boolean = false,
    val changedLanguage: Boolean = false,
    val changedRecordingOptOut: Boolean = false,
    val changedTrack: Boolean = false,
    val changedIsNew: Boolean = false,
    val changedIsCanceled: Boolean = false,
)
```

**Review by cketti — DISMISSED** [review]

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/repositories/AppRepositorySessionsTest.kt (hunk 3 above)
Comment (johnjohndoe) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/repositories/AppRepositorySessionsTest.kt:119]: :x: I like to keep the helper function to at least not repeat the `url` statement over and over.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/models/Session.kt (hunk 6 above)
Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/models/Session.kt:122]: :+1: Agreed. Very much. I noted it down for a follow-up refactoring.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/repositories/AppRepositorySessionsTest.kt (hunk 3 above)
Comment (cketti) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/repositories/AppRepositorySessionsTest.kt:119]: The value for the `url` parameter is the same as the default value. So it's not really doing anything.

File: app/src/main/java/nerd/tuxmobil/fahrplan/congress/models/Session.kt (hunk 1 above)
Comment (johnjohndoe) [reply app/src/main/java/nerd/tuxmobil/fahrplan/congress/models/Session.kt]: :heavy_check_mark: Changed now.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/models/SessionTest.kt (hunk 2 above)
Comment (johnjohndoe) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/models/SessionTest.kt]: :heavy_check_mark: Removed.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/sharing/SimpleSessionFormatTest.kt (hunk 4 above)
Comment (johnjohndoe) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/sharing/SimpleSessionFormatTest.kt]: :heavy_check_mark: Applied formatting.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/sharing/SimpleSessionFormatTest.kt (hunk 5 above)
Comment (johnjohndoe) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/sharing/SimpleSessionFormatTest.kt]: :heavy_check_mark: Applied formatting.

Yes, but I haven't taken the time to look into it.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/serialization/ScheduleChangesTest.kt
Comment (johnjohndoe) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/serialization/ScheduleChangesTest.kt:1]: I prefer to go with the `ParameterizedTest` - at least the assertion part is not duplicated in each scenario.

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/repositories/AppRepositorySessionsTest.kt (hunk 3 above)
Comment (johnjohndoe) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/repositories/AppRepositorySessionsTest.kt]: Interesting. Simplified now. :heavy_check_mark:

**Review by cketti — APPROVED** [review]

File: app/src/test/java/nerd/tuxmobil/fahrplan/congress/serialization/ScheduleChangesTest.kt
Comment (cketti) [reply app/src/test/java/nerd/tuxmobil/fahrplan/congress/serialization/ScheduleChangesTest.kt:1]: No.

**Pull request merged**